### PR TITLE
clustermesh: enable per-cluster RBAC in etcd server

### PR DESCRIPTION
**Pull request by giorio94**
This PR reworks how the cilium agents and the cluster-mesh API server are authorized against the clustermesh etcd server, configuring per-cluster RBAC policies, to improve the overall security and provide better isolation.

In particular:

* The clustermesh API server is switched to use an ad-hoc user rather than root.
* The agents on each cluster use a different user against the remote etcd servers. A new logic is introduced in the clustermesh API server, and it is responsible for the creation of the etcd users, based on the configuration provided. This per-cluster users operating mode is disabled by default, and it can be enabled through a dedicated helm values flag. An intermediate configuration is additionally provided to allow switching from one mode to another with no disruption.

Additional details are provided in the respective commits.

@reviewers: this is the first time I use the Hive framework. Feel free to let me know if anything should be implemented differently.

```release-note
clustermesh: enable per-cluster RBAC in etcd server
```

### CODE REVIEW

**Review by qmonnet — APPROVED** [review]
Sorry, I didn't spend the time to look at the code in details - `Documentation/` changes are good.

By the way, should the feature get documented?

**Comment by giorio94** [comment]
> By the way, should the feature get documented?

I would personally say that the documentation is blocked until the cilium CLI is migrated to use helm. The documentation is based on the CLI, but it currently uses hard-coded manifests, which make impossible to enable this feature.

**Review by jrajahalme — COMMENTED** [review]
How do we add new clusters to the clustermesh with the auth mode `cluster`? Seems all clusters need to have the (new) user of the new cluster added, how does that work?

**Comment by giorio94** [comment]
> How do we add new clusters to the clustermesh with the auth mode `cluster`? Seems all clusters need to have the (new) user of the new cluster added, how does that work?

I agree that all clusters need to have the new user added, but IMHO that is not much different compared to now, since the new etcd configuration needs anyhow to be set on all clusters. In this case, users are automatically synchronized in etcd by the clustermesh API server based on the content of a ConfigMap, which is updated through helm when a new cluster is listed. There is a transient period in which the new user might not be yet present (depending on the ordering of the operations), but the connection will eventually be established once it gets created.

**Comment by giorio94** [comment]
The ConformanceIngress test failed due to https://github.com/cilium/cilium/issues/24347. Retriggered

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml [hunk 1]
```diff
@@ -48,18 +48,15 @@ spec:
         - |
           rm -rf /var/run/etcd/*;
           /usr/local/bin/etcd --data-dir=/var/run/etcd --name=clustermesh-apiserver --listen-client-urls=http://127.0.0.1:2379 --advertise-client-urls=http://127.0.0.1:2379 --initial-cluster-token=clustermesh-apiserver --initial-cluster-state=new --auto-compaction-retention=1 &
```
Comment (squeed) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:53]: I've always thought this initContainer was awkward. While you're here, do you want to move this to a postStart hook?

File: pkg/kvstore/etcd.go [hunk 2]
```diff
@@ -1716,6 +1716,49 @@ func (e *etcdClient) ListAndWatch(ctx context.Context, name, prefix string, chan
...
+			scopedLog.Debug("User already exists")
+		} else {
+			return err
+		}
+	}
+
+	for _, role := range roles {
+		scopedLog.WithField(FieldRole, role).Debug("Granting role to user")
+
+		_, err := e.client.Auth.UserGrantRole(ctx, name, role)
```
Comment (squeed) [inline pkg/kvstore/etcd.go:1747]: Does this return an error if the user already has this role?

File: pkg/kvstore/client.go [hunk 3]
```diff
@@ -102,3 +107,40 @@ func Connected() <-chan struct{} {
...
+		var wg sync.WaitGroup
+
+		lc.Append(hive.Hook{
+			OnStart: func(hive.HookContext) error {
+				wg.Add(1)
+				go func() {
+					select {
+					case <-defaultClientSet:
+						resolver.Resolve(defaultClient)
+					case <-stop:
```
Comment (squeed) [inline pkg/kvstore/client.go]: You should also handle the case of hive.HookContext being closed, since we can abort startup.

File: clustermesh-apiserver/users_mgmt.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	}()
+
+	return nil
+}
+
+func (us *usersManager) Stop(hive.HookContext) error {
+	log.WithField(logfields.Path, us.ClusterUsersConfigPath).
+		Info("Stopping managing etcd users based on configuration")
+
+	us.manager.RemoveAllAndWait()
```
Comment (squeed) [inline clustermesh-apiserver/users_mgmt.go:130]: Why is this necessary?

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml [hunk 5]
```diff
@@ -167,6 +174,11 @@ spec:
         - name: etcd-admin-client
           mountPath: /var/lib/cilium/etcd-secrets
           readOnly: true
+        {{- if ne .Values.clustermesh.apiserver.tls.authMode "legacy" }}
```
Comment (squeed) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:177]: Any reason not to always mount this?

Comment (squeed) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:177]: (The configMap Volume could just be `optional: true`)

**Review by squeed — COMMENTED** [review]
Really nice! Just a few questions, and a chance for maybe one more cleanup, but looks really good.

File: pkg/kvstore/etcd.go (hunk 2 above)
Comment (giorio94) [reply pkg/kvstore/etcd.go:1747]: It does not: https://github.com/etcd-io/etcd/blob/main/server/auth/store.go#L542-L551

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml (hunk 1 above)
Comment (giorio94) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:53]: I agree that it is awkward, but I don't think that is possible, since using a postStart hook would leave a time window with no RBACs enforcement during startup. It currently does not happen since the init container configures etcd to listen only on localhost, while it is exposed publicly when started by the main container. IMHO, the best solution would be if etcd provided the possibility to RBAC parameters through a config file, but that is not possible as of today.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml (hunk 5 above)
Comment (giorio94) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:177]: No particular reasons, it just seemed to me a bit cleaner to not mount it if not needed, as a restart of the pod would still be caused upon authMode change due to the different parameters. I'm open to change it if you feel always mounting it is better.

File: clustermesh-apiserver/users_mgmt.go (hunk 4 above)
Comment (giorio94) [reply clustermesh-apiserver/users_mgmt.go:130]: It ensures that the controller which takes care of synchronizing the users gets stopped when Stop is called. And it waits for it to terminate as per the second guideline here: https://docs.cilium.io/en/latest/contributing/development/hive/#guidelines (if I got your question correctly).

File: pkg/kvstore/client.go (hunk 3 above)
Comment (giorio94) [reply pkg/kvstore/client.go]: I've double checked this, but the `OnStart` context [gets canceled once the Start hooks terminate](https://github.com/cilium/cilium/blob/master/pkg/hive/lifecycle.go#L86-L90), and at that point the etcd client might not yet haven been configured (AFAIU, it [should be used if `OnStart` blocks](https://github.com/cilium/cilium/blob/master/pkg/hive/lifecycle.go#L17-L21), but in this case that is done in a separate goroutine). In case startup is aborted, though, the corresponding `OnStop` method should be called, and that would cause the `stop` channel to be closed and the goroutine to terminate.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml (hunk 1 above)
Comment (squeed) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:53]: Yeah, makes sense. Annoying.

File: pkg/kvstore/client.go (hunk 3 above)
Comment (squeed) [reply pkg/kvstore/client.go]: oops, missed that extra `go`, nevermind!

File: clustermesh-apiserver/users_mgmt.go (hunk 4 above)
Comment (squeed) [reply clustermesh-apiserver/users_mgmt.go:130]: Aha, I mis-parsed this as removing all etcd users :-).

**Review by squeed — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]

File: pkg/kvstore/client.go [hunk 6]
```diff
@@ -102,3 +107,40 @@ func Connected() <-chan struct{} {
 	}(c)
 	return c
 }
+
+// GlobalUserMgmtClientPromiseCell provides a promise returning the global kvstore client to perform users
+// management operations, once it has been initialized. Note: client initialization must be handled separately.
+var GlobalUserMgmtClientPromiseCell = cell.Module(
```
Comment (tklauser) [inline pkg/kvstore/client.go]: It looks like most cells currently in the code base live in their a dedicated `cell.go` file. That way they're easier to discover. So maybe move this to `pkg/kvstore/cell.go`?

**Review by tklauser — APPROVED** [review]
LGTM. One minor nit regarding the file in which the `GlobalUserMgmtClientPromiseCell` is set up. Probably not worth a respin just for that. May can be addressed in a follow-up PR.

**Comment by giorio94** [comment]
The first force-push rebased onto master to fix the conflict, while the second one moved the promise logic to a separate `pkg/kvstore/cell.go` file as per Tobias's suggestion.
